### DISCUSSION TITLE
How do I fetch a URL on startup?

### DISCUSSION THREAD

**Discussion by kaihendry** [serverless/serverless#10398] [Q&A]
I want to fetch a resource needed for my app on startup just the once. How does one achieve that?

This doesn't work:

```
const axios = require("axios");

async function get_jwks() {
  const url = "http://worldtimeapi.org/api/timezone/Asia/Singapore";
  const res = await axios(url);
  return res.data.datetime;
}

const jwks_on_load = get_jwks();

module.exports.hello = async (event) => {
  return {
    statusCode: 200,
    body: JSON.stringify({ date: jwks_on_load, foo: "bar" }),
  };
};
```

nor this 

```
const axios = require("axios");

var t = "";

async function get_jwks() {
  const url = "http://worldtimeapi.org/api/timezone/Asia/Singapore";
  const res = await axios(url);
  t = res.data.datetime;
}

get_jwks();

module.exports.hello = async (event) => {
  console.log("here", t);
  return {
    statusCode: 200,
    body: JSON.stringify({ date: t, foo: "bar" }),
  };
};

```

What am I missing? :sweat_smile:

**Comment by pgrzesik** [comment]
You're missing an `await` on the `get_jwks` result - it returns a promise, not an actual value.

**Reply by kaihendry** [reply]
When I try putting an await there, I get a `"errorMessage":"await is only valid in async functions`, so I'm a little puzzled.

**Reply by pgrzesik** [reply]
Top level `await` is not supported in Lambda runtime, but you can await the result in the function body, so use `const jwks_on_load = get_jwks()` and then `await jwks_on_load` in function.

**Reply by kaihendry** [reply]
But I don't want time to update everytime I invoke the function, I want it to "keep warm", if that makes any sense?

```
const axios = require("axios");

var t = "";

async function get_jwks() {
  const url = "http://worldtimeapi.org/api/timezone/Asia/Singapore";
  const res = await axios(url);
  t = res.data.datetime;
}

const jwks_on_load = get_jwks();

module.exports.hello = async (event) => {
  await get_jwks();
  console.log("here", t);
  return {
    statusCode: 200,
    body: JSON.stringify({ date: t, foo: "bar" }),
  };
};
```

**Reply by pgrzesik** [reply]
As suggested, you should not await a new call of the function, but the cached promise, so do `await jwks_on_load`, not `await get_jwks()`. Alternatively, you can also first inspect if `t` is set and just don't do anything if it is

**Reply by kaihendry** [reply]
I think I was getting terribly confused by `serverless offline` since it doesn't behave the same as lambda in this regard! :laughing:

**Reply by pgrzesik** [reply]
That's correct @kaihendry - `offline` doesn't always behave the same way, which might be tricky to debug in some cases.
